### PR TITLE
Refactor our analyzers to support multiple diagnostic IDs.

### LINKED ISSUE #177: Ability to associate several rules to one single DiagnosticAnalyzer
We start having some rules which closely relate to each other relate either because they are raising very similar issues (see #165) or because they depends upon the same engine (see #176 Symbolic execution engine should be executed once and only once).

In both cases, the capability to have one DiagnoticAnalyzer covering several rules would be a perfect solution.

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #177

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs [hunk 1]
```diff
@@ -323,72 +367,53 @@ private static ValidationFailure IsSimpleString(int formatStringItemsCount, int
...
+            public static readonly ValidationFailure SimpleString =
+                new ValidationFailure("Remove this formatting call and simply use the input string.");
+            public static readonly ValidationFailure UnknownError =
+                new ValidationFailure("Invalid string format, the format string is invalid and is likely to throw at runtime.");
+            public static readonly ValidationFailure MissingFormatItemIndex =
+                new ValidationFailure("The format string might be wrong, the following item indexes are missing: ");
+            public static readonly ValidationFailure UnusedFormatArguments =
+                new ValidationFailure("The format string might be wrong, the following arguments are unused: ");
+
+            private string message;
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs]: MAJOR Make 'message' 'readonly'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2933)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [StringFormatValidator.cs#L209](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs#L209): Refactor this code to not nest more than 3 control flow statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS134)
1. ![MAJOR][MAJOR] [StringFormatValidator.cs#L308](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs#L308): Remove this unused '_' local variable. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1481)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs (hunk 1 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs]: MAJOR Make 'message' 'readonly'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2933)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [StringFormatValidator.cs#L209](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs#L209): Refactor this code to not nest more than 3 control flow statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS134)
1. ![MAJOR][MAJOR] [StringFormatValidator.cs#L308](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs#L308): Remove this unused '_' local variable. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1481)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [StringFormatValidator.cs#L209](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs#L209): Refactor this code to not nest more than 3 control flow statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS134)
1. ![MAJOR][MAJOR] [StringFormatValidator.cs#L308](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs#L308): Remove this unused '_' local variable. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1481)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnaryPrefixOperatorRepeated.cs [hunk 2]
```diff
@@ -69,7 +70,7 @@ protected override void Initialize(SonarAnalysisContext context)
                     }
 
                     var errorLocation = new TextSpan(topLevelUnary.SpanStart, lastUnary.OperatorToken.Span.End - topLevelUnary.SpanStart);
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnaryPrefixOperatorRepeated.cs]: BLOCKER 'topLevelUnary' is null on at least one execution path. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2259)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs (hunk 1 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs]: MAJOR Make 'message' 'readonly'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2933)

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 6 major
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 7 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [BooleanLiteralUnnecessary.cs#L117](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/BooleanLiteralUnnecessary.cs#L117): Change this condition so that it does not always evaluate to 'true'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2583)
1. ![MAJOR][MAJOR] [ConditionalSimplification.cs#L309](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionalSimplification.cs#L309): Refactor this code to not nest more than 3 control flow statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS134)
1. ![MAJOR][MAJOR] [ConditionalSimplification.cs#L314](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionalSimplification.cs#L314): Refactor this code to not nest more than 3 control flow statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS134)
1. ![MAJOR][MAJOR] [InheritedCollidingInterfaceMembers.cs#L109](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InheritedCollidingInterfaceMembers.cs#L109): Refactor this code to not nest more than 3 control flow statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS134)
1. ![MAJOR][MAJOR] [StringFormatValidator.cs#L209](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs#L209): Refactor this code to not nest more than 3 control flow statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS134)
1. ![MAJOR][MAJOR] [StringFormatValidator.cs#L308](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs#L308): Remove this unused '_' local variable. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1481)
1. ![INFO][INFO] [TokenTypeAnalyzerBase.cs#L225](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/Utilities/TokenTypeAnalyzerBase.cs#L225): Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

[BLOCKER]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER'
[CRITICAL]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 6 major
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 7 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [BooleanLiteralUnnecessary.cs#L117](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/BooleanLiteralUnnecessary.cs#L117): Change this condition so that it does not always evaluate to 'true'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2583)
1. ![MAJOR][MAJOR] [ConditionalSimplification.cs#L309](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionalSimplification.cs#L309): Refactor this code to not nest more than 3 control flow statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS134)
1. ![MAJOR][MAJOR] [ConditionalSimplification.cs#L314](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionalSimplification.cs#L314): Refactor this code to not nest more than 3 control flow statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS134)
1. ![MAJOR][MAJOR] [InheritedCollidingInterfaceMembers.cs#L109](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InheritedCollidingInterfaceMembers.cs#L109): Refactor this code to not nest more than 3 control flow statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS134)
1. ![MAJOR][MAJOR] [StringFormatValidator.cs#L209](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs#L209): Refactor this code to not nest more than 3 control flow statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS134)
1. ![MAJOR][MAJOR] [StringFormatValidator.cs#L308](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs#L308): Remove this unused '_' local variable. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1481)
1. ![INFO][INFO] [TokenTypeAnalyzerBase.cs#L225](https://github.com/SonarSource/sonar-csharp/blob/054adc612306f1135ce31432f1197096ebd0c899/sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/Utilities/TokenTypeAnalyzerBase.cs#L225): Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

[BLOCKER]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER'
[CRITICAL]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Comment by Evangelink** [comment]
@valhristov Ping review!

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs (hunk 1 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs]: MAJOR Make 'message' 'readonly'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2933)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [StringFormatValidator.cs#L209](https://github.com/SonarSource/sonar-csharp/blob/19a3789d880a38ea57bcec4b07a14463dc7ec699/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs#L209): Refactor this code to not nest more than 3 control flow statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS134)
1. ![MAJOR][MAJOR] [StringFormatValidator.cs#L308](https://github.com/SonarSource/sonar-csharp/blob/19a3789d880a38ea57bcec4b07a14463dc7ec699/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs#L308): Remove this unused '_' local variable. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1481)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs (hunk 1 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs]: MAJOR Make 'message' 'readonly'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2933)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [StringFormatValidator.cs#L209](https://github.com/SonarSource/sonar-csharp/blob/19a3789d880a38ea57bcec4b07a14463dc7ec699/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs#L209): Refactor this code to not nest more than 3 control flow statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS134)
1. ![MAJOR][MAJOR] [StringFormatValidator.cs#L308](https://github.com/SonarSource/sonar-csharp/blob/19a3789d880a38ea57bcec4b07a14463dc7ec699/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidator.cs#L308): Remove this unused '_' local variable. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1481)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Pull request merged**